packages/loader-react/documentation/SkeletonLoaderExample.tsx

```diff
@@ -55,36 +55,40 @@ export const SkeletonLoaderExample: FC<ExampleComponentProps> = () => {
 };
 
 export const skeletonLoaderExampleCode: CodeExample = () => {
-    return `<SkeletonAnimation
-    role="none presentation" // Kun for eksempelet, ellers blir det mange alerts på siden
->
-    <div style={{ display: "flex", flexDirection: "column", gap: 40 }}>
-        <SkeletonElement width={350} height={32} />
-        <SkeletonInput />
-        <SkeletonTextArea />
-        <SkeletonCheckboxGroup checkboxes={3} />
-        <SkeletonRadioButtonGroup radioButtons={3} />
-        <SkeletonButton width={225} />
-        <SkeletonTable>
-            <SkeletonTableHeader>
-                <div style={{ width: 215 }}>
-                    <SkeletonElement width={115} height={16} />
-                </div>
-                <div style={{ width: 115 }}>
-                    <SkeletonElement width={115} height={16} />
-                </div>
-            </SkeletonTableHeader>
-            {Array.from(Array(5)).map((_, index) => (
-                <SkeletonTableRow key={index}>
+    return `
+{/* Vis skeleton loader, og innholdet som lastes inn, i en live region for å annonsere for skjermlesere. */}
+<div role="status" aria-busy="true">
```

Comment:
Dette er visstnok ikke så rett fram likevel. En live region som i praksis går over hele siden gir potensielt andre fallgruver, og [det er tilsynelatende ikke noen konsensus](https://adrianroselli.com/2020/11/more-accessible-skeletons.html) om hvordan skeletons best løses med tanke på UU i utgangspunktet.

Reply:
Cc @joms @piofinn, dette er vanskelig 😅

Reply:
**tl;dr:** Skeleton er to ting, loader og placeholder. Bør de være mer separate enn i dag?

Jeg har brukt litt tid på å lese om og tenke på dette, og det jeg har kommet frem til så langt er at mangelen på konsensus rundt dette er at _skeleton loaders er to ting_:

- Skeleton loaders indikerer _at_ innhold lastes inn på siden, altså en indikasjon på applikasjonstilstand.
- Skeleton loaders er en abstrakt representasjon av _hvilket innhold_ som kommer til å dukke opp på siden, altså en placeholder

Kravene til disse to tingene er forskjellige, og derfor blir det ikke rett frem hva slags rolle og egenskaper skeleton loaderen skal ha. Hvis vi klarer å separere disse to rollene litt – gjerne innenfor samme pakke – blir det kanskje enklere å jobbe med?

Så hjelper det jo ikke at det ikke er definert ordentlige retningslinjer i WCAG for innlasting av innhold. Det nærmeste jeg har kommet er [dette forslaget](https://www.w3.org/WAI/GL/wiki/Notification_of_Loading/Busy) fra 2016 til et nytt punkt. Der er "konklusjonen" bruk av live-regioner for å gjøre brukeren oppmerksom på endringer på siden.

Likevel, hvis vi ser på de to rollene kan vi prøve å komme frem til noe:

#### Indikasjon på at noe lastes inn

Så vidt jeg kan se er det dette som egentlig betyr mest, og derfor har flest krav knyttet til seg. I forslaget jeg nevnte over vil man kunne si at man bør **si fra via en live region** at innholdet er i ferd med å lastes inn, og om det eventuelt feiler eller blir lastet inn. Dette er ekstra viktig hvis innholdet endrer seg etter at siden er lastet inn første gang (initial load).

Her kan det hende at det også vil være krav til kontrast, avhengig av hvordan man tolker kravene under punkt 1.4 i WCAG. [Denne kommentaren](https://github.com/w3c/wcag/issues/2048#issuecomment-1516806427) argumenterer godt for at det vil være et krav.

#### Representasjon av innholdet som lastes inn (placeholder)

Her har jeg egentlig ikke klart å finne så mange krav, og vet ikke helt om det burde være så mange heller. Oppgaven til placeholderen er jo egentlig bare å indikere løst hva man kan forvente å se når innholdet er lastet inn, samt å unngå layout shift. For meg har dette ikke så mye semantisk mening, og bør kanskje derfor heller ikke ha noen rolle?

### Forslag

Ut fra dette tenker jeg at indikasjonen på innlasting, som altså har krav knyttet til seg, behandles som en egen greie/et eget konsept. Denne kan gjerne være en del av Skeleton-pakka, men det viktigste er kanskje at den beskrives som (en del av) et mønster. Dette vil tilsvare `<span>Loading</span>` i artikkelen @wkillerud lenka til. Er litt mer usikker på hvordan man skal løse det _visuelt_ hvis man vil fjerne denne rollen fra selve skeleton-komponentene, men det kan utforskes:

![Loading bar](https://github.com/fremtind/jokul/assets/25739615/3a9a74fc-3b12-4654-bf2c-f205b5aa7fbb)
![Spinner](https://github.com/fremtind/jokul/assets/25739615/0ee07392-176a-4cc1-bcb3-a1128da199a1)

Reply:
Eller så kanskje det bare holder å rendre ut en span med [rollen `status`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/status_role) inne i `SkeletonAnimation` (med prop for teksten) og så ikke tenke mer på det? 😅 F.eks. for skissene over:

```html
<span class="jkl-sr-only" role="status">Laster inn fakturaoversikt</span>
```

Reply:
Det vil ikke hjelpe en skjermleser med å fange opp selve innholdet når det blir lastet inn.

Reply:
Nå vil vi vel ikke nødvendigvis at skjermleser skal lese opp alt som lastes inn når det er lastet inn. Live regions er jo egentlig til for ting som endrer seg ute av kontekst av det brukeren gjør, som stillingen i en fotballkam, kursen til en aksje, eller en viktig advarsel om en hendelse.

Sånn sett er det kanskje ikke nødvendig med _noe_ så lenge lastingen skjer i forbindelse med første innlasting av en side? Det er jo ingen indikasjoner for rent HTML-innhold som ikke er rendret ennå på en treg forbindelse, for eksempel: Det er bare ikke tilgjengelig i DOMet ennå. Eller litt mer visuelt, er det noen konseptuell forskjell mellom de to opplevelsene under ved innlasting av en side?

https://github.com/fremtind/jokul/assets/25739615/480dbb46-47fd-4ed2-aca2-270b4c7a7f99


https://github.com/fremtind/jokul/assets/25739615/974527e1-d521-4e82-9899-08c900ef12e2

I det første tilfellet ville man vel godtatt at det ikke var noen tilrettelegging for skjermleser (annet enn eventuelt en indikasjon på at siden fortsatt laster inn), så hvorfor ikke i det andre tilfellet? Om noe er kanskje det man trenger en indkasjon på at innholdet _er ferdig innlastet_ mer enn at det _lastes inn_?

Reply:
Hm, kanskje. Her kan jeg ikke nok om brukeropplevelsen, rett og slett. Men du har nok rett i at det ikke er noen forskjell mellom de to opplevelsene.